browser/src/core/geometry.ts

```diff
@@ -0,0 +1,226 @@
+
+	public get pArea(): number { return this.pWidth * this.pHeight; }
+	public get pCenter(): number[] { return [(this.pX1 + this.pX2) / 2, (this.pY1 + this.pY2) / 2]; }
+
+	public pToArray(): number[] { return [this.pX1, this.pY1, this.pWidth, this.pHeight]; }
+
+	// Pixel checkers for coordinate match.
+	public pContainsPoint (point: number[]): boolean { return (Math.round(point[0]) >= this.pX1 && Math.round(point[0]) <= this.pX2 && Math.round(point[1]) >= this.pY1 && Math.round(point[1]) <= this.pY2); }
+	public pContainsX (x: number): boolean { return (Math.round(x) >= this.pX1 && Math.round(x) <= this.pX2); }
+	public pContainsY (y: number): boolean { return (Math.round(y) >= this.pY1 && Math.round(y) <= this.pY2); }
+	public pEquals(rectangle: Array<number>): boolean { return this.pX1 === Math.round(rectangle[0]) && this.pY1 === Math.round(rectangle[1]) && this.pWidth === Math.round(rectangle[2]) && this.pHeight === Math.round(rectangle[3]); }
+
+	public pMoveTo (point: number[]): void { this._x1 = Math.round(point[0] * app.pixelsToTwips); this._y1 = Math.round(point[1] * app.pixelsToTwips); }
+	public pMoveBy (point: number[]): void { this._x1 += Math.round(point[0] * app.pixelsToTwips); this._y1 += Math.round(point[1] * app.pixelsToTwips); }
+
+	// CSS pixel.
+	public get cX1(): number { return Math.round(this._x1 * app.twipsToPixels / app.dpiScale); }
+	public set cX1 (x1: number) { this._x1 = Math.round(x1 * app.dpiScale * app.pixelsToTwips); }
+
+	public get cY1(): number { return Math.round(this._y1 * app.twipsToPixels / app.dpiScale); }
+	public set cY1 (y1: number) { this._y1 = Math.round(y1 * app.dpiScale * app.pixelsToTwips); }
+
+	public get cX2(): number { return this.cX1 + this.cWidth; }
+	public set cX2 (x2: number) { this._width = Math.round(x2 * app.dpiScale * app.pixelsToTwips); }
+
+	public get cY2(): number { return this.cY1 + this.cHeight; }
+	public set cY2 (y2: number) { this._height = Math.round(y2 * app.dpiScale * app.pixelsToTwips); }
+
+	public get cWidth(): number { return Math.round(this._width * app.twipsToPixels / app.dpiScale); }
+	public set cWidth (width: number) { this._width = Math.round(width * app.dpiScale * app.pixelsToTwips); }
+
+	public get cHeight(): number { return Math.round(this._height * app.twipsToPixels / app.dpiScale); }
+	public set cHeight (height: number) { this._height = Math.round(height * app.dpiScale * app.pixelsToTwips); }
+
+	public get cArea(): number { return this.cWidth * this.cHeight; }
+	public get cCenter(): number[] { return [(this.cX1 + this.cX2) / 2, (this.cY1 + this.cY2) / 2]; }
+
+	public cToArray(): number[] { return [this.cX1, this.cY1, this.cWidth, this.cHeight]; }
+
+	// CSS pixel checkers for coordinate match.
+	public cContainsPoint (point: number[]): boolean { return (Math.round(point[0]) >= this.cX1 && Math.round(point[0]) <= this.cX2 && Math.round(point[1]) >= this.cY1 && Math.round(point[1]) <= this.cY2); }
```

Comment:
we have SimplePoint class, why do we pass regular array? Wouldn't be better to pass SimplePoint, so we will be sure inside it will use correct value?

with untyped array of just number we be still affected by programmer mistakes in conversions

Reply:
I wanted not to enforce it until we have consistent instances accross app. I added this thought in the comment at the start of the file.

Reply:
Doesn't that mean we will need 2 times iterate over the full code?
1. use this Simple classes but with plain "numbers" and "arrays"
2. after all is completed, again change arrays into new types

I believe source of some of our problems is that we can use any incompatible numbers and still get "some" result.

So I would recommend to:
- on input allow plain numbers as you have, it should be always TWIP as you say
- on output (all cX(), pX(), x() methods) should return typed value. eg `type CssPixel = number;`
  it makes readability better because you don't need to look at comment to understand which unit it is,
  also will help to warn about incorrect usage

Thanks to that, if we switch code to use of new class, it will be done in single step

Reply:
But sure, it is good enough now to at least have this version

Reply:
> Doesn't that mean we will need 2 times iterate over the full code?
   I don't think the we spend on this will be longer. There is no consensus yet between class types. I understand your concerns, but this way seems safer to me.

> we have SimplePoint class, why do we pass regular array? Wouldn't be better to pass SimplePoint, so we will be sure inside it will use correct value?
   Where there is a ts file, a number array or point class instance doesn't matter. Type is checked. If it is a JS file, type doesn't matter this time, it's not checked. Same thing with or without point class.

> on output (all cX(), pX(), x() methods) should return typed value. eg type CssPixel = number;
> it makes readability better because you don't need to look at comment to understand which unit it is,
> also will help to warn about incorrect usage
   This is a good idea:) But i still need to check this. Seems that CSSPixels and numbers can be added but we need to be sure that we won't get unnecessary linting warnings when  we add +1 to a CSSPixel type. I will check this. Let me note and we will return to this one.